url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/canvas/mxSvgCanvas.java

```diff
@@ -0,0 +1,774 @@
+
+		if (mxUtils.isTrue(style, mxConstants.STYLE_SHADOW, false)
+				&& !fillColor.equals("none"))
+		{
+			shadowElement = (Element) bg.cloneNode(true);
+
+			shadowElement.setAttribute("transform",
+					mxConstants.SVG_SHADOWTRANSFORM);
+			shadowElement.setAttribute("fill", mxConstants.W3C_SHADOWCOLOR);
+			shadowElement.setAttribute("stroke", mxConstants.W3C_SHADOWCOLOR);
+			shadowElement.setAttribute("stroke-width", String
+					.valueOf(strokeWidth));
+
+			appendSvgElement(shadowElement);
+		}
+
+		// Applies rotation
+		double rotation = mxUtils.getDouble(style, mxConstants.STYLE_ROTATION);
+
+		if (rotation != 0)
+		{
+			int cx = x + w / 2;
+			int cy = y + h / 2;
+
+			elem.setAttribute("transform", "rotate(" + rotation + "," + cx
+					+ "," + cy + ")");
+
+			if (shadowElement != null)
+			{
+				shadowElement.setAttribute("transform", "rotate(" + rotation
+						+ "," + cx + "," + cy + ") "
+						+ mxConstants.SVG_SHADOWTRANSFORM);
+			}
+		}
+
+		// Applies opacity
+		float opacity = mxUtils.getFloat(style, mxConstants.STYLE_OPACITY, 100);
+
+		if (opacity != 100)
+		{
+			String value = String.valueOf(opacity / 100);
+			elem.setAttribute("fill-opacity", value);
+			elem.setAttribute("stroke-opacity", value);
+
+			if (shadowElement != null)
+			{
+				shadowElement.setAttribute("fill-opacity", value);
+				shadowElement.setAttribute("stroke-opacity", value);
+			}
+		}
+
+		if (mxUtils.isTrue(style, mxConstants.STYLE_DASHED))
+		{
+			elem.setAttribute("stroke-dasharray", "3, 3");
+		}
+
+		appendSvgElement(elem);
+
+		return elem;
+	}
+
+	/**
+	 * Draws the given lines as segments between all points of the given list
+	 * of mxPoints.
+	 * 
+	 * @param pts List of points that define the line.
+	 * @param style Style to be used for painting the line.
+	 */
+	public Element drawLine(List<mxPoint> pts, Map<String, Object> style)
+	{
+		Element group = document.createElement("g");
+		Element path = document.createElement("path");
+
+		String strokeColor = mxUtils.getString(style,
+				mxConstants.STYLE_STROKECOLOR);
+		float tmpStroke = (mxUtils.getFloat(style,
+				mxConstants.STYLE_STROKEWIDTH, 1));
+		float strokeWidth = (float) (tmpStroke * scale);
+
+		if (strokeColor != null && strokeWidth > 0)
+		{
+			// Draws the start marker
+			Object marker = style.get(mxConstants.STYLE_STARTARROW);
+
+			mxPoint pt = pts.get(1);
+			mxPoint p0 = pts.get(0);
+			mxPoint offset = null;
+
+			if (marker != null)
+			{
+				float size = (mxUtils.getFloat(style,
+						mxConstants.STYLE_STARTSIZE,
+						mxConstants.DEFAULT_MARKERSIZE));
+				offset = drawMarker(group, marker, pt, p0, size, tmpStroke,
+						strokeColor);
+			}
+			else
+			{
+				double dx = pt.getX() - p0.getX();
+				double dy = pt.getY() - p0.getY();
+
+				double dist = Math.max(1, Math.sqrt(dx * dx + dy * dy));
+				double nx = dx * strokeWidth / dist;
+				double ny = dy * strokeWidth / dist;
+
+				offset = new mxPoint(nx / 2, ny / 2);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 522:540, 558:576.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.